### PR TITLE
feat(ingress-nginx): add backend TLS support

### PR DESCRIPTION
**Pull request by rajashish**
This commit implements support for Nginx Ingress backend TLS annotations:
- nginx.ingress.kubernetes.io/proxy-ssl-verify
- nginx.ingress.kubernetes.io/proxy-ssl-secret

Changes:
- Added ProxySSLVerifyAnnotation and ProxySSLSecretAnnotation constants.
- Implemented backendTLSFeature to parse these annotations.
- Maps 'proxy-ssl-verify: on' + 'proxy-ssl-secret' to BackendTLSPolicy validation.caCertificateRefs.
- Handles cross-namespace secret references by using the secret name (assuming same namespace as BackendTLSPolicy limits).
- Added comprehensive unit tests in backend_tls_test.go and updated converter tests.

**What type of PR is this?**
/kind feature
/kind test
**What this PR does / why we need it**:
This PR implements support for Nginx Ingress backend TLS annotations in the `ingress-nginx` provider.
It maps the following annotations to Gateway API's [BackendTLSPolicy](cci:1://file:///usr/local/google/home/aashtrek/ingress2gateway1/pkg/i2gw/providers/common/utils.go:314:0-344:1):
- `nginx.ingress.kubernetes.io/proxy-ssl-verify`: Controls whether to verify the backend server's certificate.
- `nginx.ingress.kubernetes.io/proxy-ssl-secret`: Specifies the CA certificate to use for verification.
**Why we need it**:
Users migrating from Ingress Nginx to Gateway API need to preserve their secure backend connection configurations. This change ensures that upstream TLS verification settings are correctly translated to [BackendTLSPolicy](cci:1://file:///usr/local/google/home/aashtrek/ingress2gateway1/pkg/i2gw/providers/common/utils.go:314:0-344:1), preventing security downgrades or connection failures during migration.
**Implementation Details**:
- **Logic**: Added [backendTLSFeature](cci:1://file:///usr/local/google/home/aashtrek/ingress2gateway1/pkg/i2gw/providers/ingressnginx/backend_tls.go:30:0-156:1) to parse annotations from Ingress resources.
    - Maps `proxy-ssl-verify: "on"` to a [BackendTLSPolicy](cci:1://file:///usr/local/google/home/aashtrek/ingress2gateway1/pkg/i2gw/providers/common/utils.go:314:0-344:1) with validation enabled.
    - Populates `spec.validation.caCertificateRefs` using the secret defined in `proxy-ssl-secret`.
    - Handles secret names in `namespace/name` format by extracting the name (adhering to `LocalObjectReference` constraints).
- **Tests**: Added a new test file [pkg/i2gw/providers/ingressnginx/backend_tls_test.go](cci:7://file:///usr/local/google/home/aashtrek/ingress2gateway1/pkg/i2gw/providers/ingressnginx/backend_tls_test.go:0:0-0:0) with comprehensive test cases covering:
    - Verification enabled (`ssl-verify: "on"`).
    - Secret provided with and without verification enabled.
    - Edge cases (no annotations, invalid formats).
**Which issue(s) this PR fixes**:
Fixes #
**Does this PR introduce a user-facing change?**:
```release-note
ingress-nginx: Added support for converting `proxy-ssl-verify` and `proxy-ssl-secret` annotations to [BackendTLSPolicy](cci:1://file:///usr/local/google/home/aashtrek/ingress2gateway1/pkg/i2gw/providers/common/utils.go:314:0-344:1).

### CODE REVIEW

File: pkg/i2gw/providers/ingressnginx/backend_tls.go [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+			for backendIdx, source := range backendSources {
+				if source.Ingress == nil {
+					continue
+				}
+
+				// Check for backend TLS annotations
+				proxySSLVerify := source.Ingress.Annotations[ProxySSLVerifyAnnotation]
+				proxySSLSecret := source.Ingress.Annotations[ProxySSLSecretAnnotation]
+
+				if proxySSLVerify != "on" && proxySSLSecret == "" {
```
Comment (Stevenjin8) [inline pkg/i2gw/providers/ingressnginx/backend_tls.go]: I don't think this is right. Have you tested this out on cluster? My understanding is that `backend-protocol = HTTPS` or `GRPCS` enables tls to the backend. `proxySSLVerify` just controls whether to verify the authenticity of the cert presented by the backend. Also, there are a bunch of other fields that should be included in this. https://kubernetes.github.io/ingress-nginx/user-guide/nginx-configuration/annotations/#backend-certificate-authentication

Comment (rajashish) [reply pkg/i2gw/providers/ingressnginx/backend_tls.go]: Updated the code and tested on cluster

File: pkg/i2gw/providers/ingressnginx/backend_tls.go [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+
+func backendTLSFeature(ingresses []networkingv1.Ingress, _ map[types.NamespacedName]map[string]int32, ir *providerir.ProviderIR) field.ErrorList {
+	var errList field.ErrorList
+
+	if ir.BackendTLSPolicies == nil {
+		ir.BackendTLSPolicies = make(map[types.NamespacedName]gatewayv1.BackendTLSPolicy)
+	}
+
+	for _, httpRouteContext := range ir.HTTPRoutes {
+		for ruleIdx, backendSources := range httpRouteContext.RuleBackendSources {
```
Comment (Stevenjin8) [inline pkg/i2gw/providers/ingressnginx/backend_tls.go:57]: oh we don't want to do it for every backend source, as one of them might be canary. You probably also want to work off this PR: https://github.com/kubernetes-sigs/ingress2gateway/pull/286. We should have a `getNonCanaryIngress` function somewhere

Comment (Stevenjin8) [reply pkg/i2gw/providers/ingressnginx/backend_tls.go:57]: This comment should be on line 96

File: pkg/i2gw/providers/ingressnginx/backend_tls.go [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+				proxySSLSecret := source.Ingress.Annotations[ProxySSLSecretAnnotation]
+				proxySSLName := source.Ingress.Annotations[ProxySSLNameAnnotation]
+
+				isHTTPS := backendProtocol == "HTTPS" || backendProtocol == "GRPCS"
+
+				// If not HTTPS/GRPCS, skip.
+				// Nginx only speaks TLS to backend if backend-protocol is set to HTTPS/GRPCS.
+				if !isHTTPS {
+					continue
+				}
```
Comment (Stevenjin8) [inline pkg/i2gw/providers/ingressnginx/backend_tls.go]:
```suggestion
				// If not HTTPS/GRPCS, skip.
				// Nginx only speaks TLS to backend if backend-protocol is set to HTTPS/GRPCS.
				if !( backendProtocol == "HTTPS" || backendProtocol == "GRPCS") {
					continue
				}
```

File: pkg/i2gw/providers/ingressnginx/backend_tls.go [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+
+				policyName := fmt.Sprintf("%s-backend-tls", serviceName)
+				policyKey := types.NamespacedName{Namespace: namespace, Name: policyName}
+
+				// Check if we already created a policy for this service
+				policy, exists := ir.BackendTLSPolicies[policyKey]
+				if !exists {
+					policy = common.CreateBackendTLSPolicy(namespace, policyName, serviceName)
+				}
```
Comment (Stevenjin8) [inline pkg/i2gw/providers/ingressnginx/backend_tls.go]: If there are conflicts, we need to warn.

File: pkg/i2gw/providers/ingressnginx/backend_tls.go [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+							Kind:  "Secret",
+							Name:  gatewayv1.ObjectName(secretName),
+						}}
+					} else {
+						// Verify on, no secret provided -> Implies system trust.
+						// Gateway API: "If the list is empty, the backend's certificate will be verified against the system trust store."
+						policy.Spec.Validation.CACertificateRefs = []gatewayv1.LocalObjectReference{}
+					}
+				} else {
+					// Verify off.
```
Comment (Stevenjin8) [inline pkg/i2gw/providers/ingressnginx/backend_tls.go]: In this case, we should at least emit a warning rather than doing an incorrect translation.

File: pkg/i2gw/providers/ingressnginx/backend_tls.go [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+
+				if proxySSLVerify == "on" {
+					if proxySSLSecret != "" {
+						policy.Spec.Validation.CACertificateRefs = []gatewayv1.LocalObjectReference{{
+							Group: "",
+							Kind:  "Secret",
+							Name:  gatewayv1.ObjectName(secretName),
+						}}
+					} else {
+						// Verify on, no secret provided -> Implies system trust.
```
Comment (Stevenjin8) [inline pkg/i2gw/providers/ingressnginx/backend_tls.go]: Is this true?

File: pkg/i2gw/providers/ingressnginx/backend_tls.go [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+				}
+
+				// Handle CA Certificates
+				secretName := proxySSLSecret
+				if strings.Contains(secretName, "/") {
+					parts := strings.SplitN(secretName, "/", 2)
+					if len(parts) == 2 {
+						secretName = parts[1]
+					}
+				}
```
Comment (Stevenjin8) [inline pkg/i2gw/providers/ingressnginx/backend_tls.go:172]: > This annotation expects the Secret name in the form "namespace/secretName".

https://kubernetes.github.io/ingress-nginx/user-guide/nginx-configuration/annotations/#backend-certificate-authentication

We should assume that this is true. Also, GWAPI requires that the secret live in the same namespace as the HTTPRoute. We should check this and not create the BackendTLSPolicy and emit a notification if the namespaces differ

File: pkg/i2gw/providers/ingressnginx/backend_tls.go [hunk 8]
```diff
@@ -0,0 +1,133 @@
...
+
+				if proxySSLVerify == "on" {
```
Comment (Stevenjin8) [inline pkg/i2gw/providers/ingressnginx/backend_tls.go]:
```suggestion
                // proxySSLVerify defaults to "off"
				if proxySSLVerify == "on" {
```

File: pkg/i2gw/providers/ingressnginx/backend_tls.go (hunk 4 above)
Comment (rajashish) [reply pkg/i2gw/providers/ingressnginx/backend_tls.go]: Implemented strict validation to ensure proxy-ssl-secret resides in the same namespace. Emits an error notification if a cross-namespace secret is specified.

File: pkg/i2gw/providers/ingressnginx/backend_tls.go (hunk 7 above)
Comment (rajashish) [reply pkg/i2gw/providers/ingressnginx/backend_tls.go:172]: Implemented strict validation to ensure proxy-ssl-secret resides in the same namespace. Emits an error notification if a cross-namespace secret is specified.

File: pkg/i2gw/providers/ingressnginx/backend_tls.go [hunk 9]
```diff
@@ -0,0 +1,195 @@
...
+			}
+			rule := httpRouteContext.HTTPRoute.Spec.Rules[ruleIdx]
+
+			for backendIdx, source := range backendSources {
+				if source.Ingress == nil {
+					continue
+				}
+
+				// Skip canary ingresses for BackendTLSPolicy generation
+				if source.Ingress.Annotations[CanaryAnnotation] == "true" {
```
Comment (Stevenjin8) [inline pkg/i2gw/providers/ingressnginx/backend_tls.go]: you can use getNonCanaryIngress here.

File: pkg/i2gw/providers/ingressnginx/backend_tls.go (hunk 5 above)
Comment (Stevenjin8) [reply pkg/i2gw/providers/ingressnginx/backend_tls.go]: In this case, we should at least emit a warning rather than doing an incorrect translation.

If we have verify off, then I find it hard to believe that the cert presented by the backend will be signed by the system/i.e. a well-known CA.

File: pkg/i2gw/providers/ingressnginx/annotations.go [hunk 10]
```diff
@@ -31,9 +31,16 @@ const (
...
 
 	// SSL Redirect annotation
 	SSLRedirectAnnotation = "nginx.ingress.kubernetes.io/ssl-redirect"
+
+
+	BackendProtocolAnnotation    = "nginx.ingress.kubernetes.io/backend-protocol"
+	ProxySSLVerifyAnnotation     = "nginx.ingress.kubernetes.io/proxy-ssl-verify"
+	ProxySSLSecretAnnotation     = "nginx.ingress.kubernetes.io/proxy-ssl-secret"
+	ProxySSLNameAnnotation       = "nginx.ingress.kubernetes.io/proxy-ssl-name"
+	ProxySSLServerNameAnnotation = "nginx.ingress.kubernetes.io/proxy-ssl-server-name"
```
Comment (Stevenjin8) [inline pkg/i2gw/providers/ingressnginx/annotations.go]: After doing a quick read, it seems like gateway api requires sending SNI, so `ProxySSLServerNameAnnotation` must be `on`.

File: pkg/i2gw/providers/ingressnginx/backend_tls.go [hunk 11]
```diff
@@ -0,0 +1,195 @@
...
+
+				// Map proxy-ssl-name to Hostname
+				if proxySSLName != "" {
+					policy.Spec.Validation.Hostname = gatewayv1.PreciseHostname(proxySSLName)
+				} else {
+					// If proxy-ssl-name is not set, we default to the Service DNS name
+					// because Gateway API v1 BackendTLSPolicy requires Hostname to be non-empty (no omitempty tag).
+					// This aligns with typical cluster-internal TLS where the cert often matches the service DNS.
+					dnsName := fmt.Sprintf("%s.%s.svc.cluster.local", serviceName, namespace)
+					policy.Spec.Validation.Hostname = gatewayv1.PreciseHostname(dnsName)
```
Comment (Stevenjin8) [inline pkg/i2gw/providers/ingressnginx/backend_tls.go]: is this what ingress nginx does?

File: pkg/i2gw/providers/ingressnginx/backend_tls.go [hunk 12]
```diff
@@ -0,0 +1,195 @@
...
+					}
+				} else {
+					// Verify off.
+					// We treat this as "System Trust" because Gateway API v1 BackendTLSPolicy
+					// requires at least one validation method.
+					system := gatewayv1.WellKnownCACertificatesSystem
+					policy.Spec.Validation.WellKnownCACertificates = &system
+				}
+
+				if exists {
```
Comment (Stevenjin8) [inline pkg/i2gw/providers/ingressnginx/backend_tls.go:182]: Can you do a deepEquals here

**Review by Stevenjin8 — CHANGES_REQUESTED** [review]
Still unsure about the correctness

File: pkg/i2gw/providers/ingressnginx/backend_tls.go [hunk 13]
```diff
@@ -44,27 +45,25 @@ func backendTLSFeature(ingresses []networkingv1.Ingress, _ map[types.NamespacedN
...
+				// We actually need the ingress that *owns* this backend ref source.
+				// Wait, backendSources is a list of BackendSource. Each BackendSource corresponds to *one* Ingress backend.
+				// But we are iterating over them. 
+				// Actually, `backendSources` is []BackendSource.
+				// A single BackendRef in the HTTPRoute might have come from multiple Ingress backends (e.g. merge).
+				// But usually 1:1 or N:1.
+				// The outer loop iterates `backendSources` which effectively corresponds to `rule.BackendRefs[backendIdx]`.
+				// Wait, no. `ruleBackendSources` is `[][]BackendSource`.
+				// `backendSources` is `[]BackendSource` corresponding to `rule.BackendRefs[backendIdx]`.
+				// So ALL these sources map to the SAME backend ref.
```
Comment (Stevenjin8) [inline pkg/i2gw/providers/ingressnginx/backend_tls.go]: delete

File: pkg/i2gw/providers/ingressnginx/backend_tls.go (hunk 9 above)
Comment (rajashish) [reply pkg/i2gw/providers/ingressnginx/backend_tls.go]: Done. I've refactored the logic to use 
getNonCanaryIngress(backendSources)
. This ensures we consistently use the primary (non-canary) Ingress for annotation lookups for every backend.

File: pkg/i2gw/providers/ingressnginx/backend_tls.go (hunk 5 above)
Comment (rajashish) [reply pkg/i2gw/providers/ingressnginx/backend_tls.go]: Added a WarningNotification if proxy-ssl-verify is set to 'off'. The code now defaults to System Trust in this case, as Gateway API BackendTLSPolicy requires at least one validation method

File: pkg/i2gw/providers/ingressnginx/annotations.go (hunk 10 above)
Comment (rajashish) [reply pkg/i2gw/providers/ingressnginx/annotations.go]: Correct. Gateway API BackendTLSPolicy mandates a Hostname, which implies SNI will be sent. Since we cannot technically 'disable' SNI in the generated Policy, I have added a Warning if proxy-ssl-server-name is set to 'off', informing the user that SNI will likely still be sent (via the generated Hostname).

File: pkg/i2gw/providers/ingressnginx/backend_tls.go (hunk 6 above)
Comment (rajashish) [reply pkg/i2gw/providers/ingressnginx/backend_tls.go]: Good catch. I updated the logic to respect proxy-ssl-secret even if verify is off, prioritizing the user's secret over System Trust (and creating a warning about implicit validation).

**Comment by rajashish** [comment]
> Still unsure about the correctness

Please review now

File: pkg/i2gw/providers/ingressnginx/backend_tls.go [hunk 14]
```diff
@@ -0,0 +1,210 @@
...
+	for _, httpRouteContext := range ir.HTTPRoutes {
+		for ruleIdx, backendSources := range httpRouteContext.RuleBackendSources {
+			if ruleIdx >= len(httpRouteContext.HTTPRoute.Spec.Rules) {
+				continue
+			}
+			rule := httpRouteContext.HTTPRoute.Spec.Rules[ruleIdx]
+
+			for backendIdx := range backendSources {
+				
+				// Identify the "main" ingress for this backend occurrence to check annotations.
```
Comment (Stevenjin8) [inline pkg/i2gw/providers/ingressnginx/backend_tls.go]: Could you look at the diff to get rid of extraneous comments

**Comment by Stevenjin8** [comment]
@rajashish When do you think this could get done by? It seems like we're fairly close, but this is a pretty important feature so I want to get  it merged earlier rather than later

**Comment by rajashish** [comment]
> ? It seems like we're fairly close,

I'll try to finish in next few hours

File: pkg/i2gw/providers/ingressnginx/backend_tls.go [hunk 15]
```diff
@@ -78,23 +78,55 @@ func backendTLSFeature(ingresses []networkingv1.Ingress, _ map[types.NamespacedN
...
+				// Strict Validation Rules to emit a policy
+				var validationErrors []string
+
+				if proxySSLVerify != "on" {
+					validationErrors = append(validationErrors, fmt.Sprintf("%s must be strictly 'on'", ProxySSLVerifyAnnotation))
+				}
+				if proxySSLSecret == "" {
+					validationErrors = append(validationErrors, fmt.Sprintf("%s must be provided with a trusted CA certificate", ProxySSLSecretAnnotation))
+				}
+				if proxySSLServerName == "off" { // Default is on, so blank is okay, just must not be off
```
Comment (Stevenjin8) [inline pkg/i2gw/providers/ingressnginx/backend_tls.go]: not true

https://github.com/kubernetes/ingress-nginx/blob/451747c70c6fca688e157a8329a3dd219a234fd9/internal/ingress/annotations/proxyssl/main.go#L38

File: pkg/i2gw/providers/ingressnginx/backend_tls.go [hunk 16]
```diff
@@ -0,0 +1,189 @@
...
+				}
+
+				policyName := fmt.Sprintf("%s-backend-tls", serviceName)
+				policyKey := types.NamespacedName{Namespace: namespace, Name: policyName}
+
+				// Check if we already created a policy for this service
+				existingPolicy, exists := ir.BackendTLSPolicies[policyKey]
+				var policy gatewayv1.BackendTLSPolicy
+				if exists {
+					policy = existingPolicy
```
Comment (Stevenjin8) [inline pkg/i2gw/providers/ingressnginx/backend_tls.go]: can we do a deep copy here to not have to deal with point shenanigans

**Review by Stevenjin8 — CHANGES_REQUESTED** [review]
Some very minor changes. LGTM otherwise

File: pkg/i2gw/providers/ingressnginx/backend_tls.go [hunk 17]
```diff
@@ -0,0 +1,189 @@
...
+							primaryIngress,
+						)
+					}
+					// If exists, we keep the existing one (first wins strategy)
+					continue
+				}
+
+				ir.BackendTLSPolicies[policyKey] = policy
+				
+				break
```
Comment (bexxmodd) [inline pkg/i2gw/providers/ingressnginx/backend_tls.go]: Why are we breaking here? if I understood it correctly,  If an Ingress rule has multiple backends (e.g., in a weighted load-balancing or canary setup), only the first service will get a BackendTLSPolicy?

File: pkg/i2gw/providers/ingressnginx/utils.go [hunk 18]
```diff
@@ -31,5 +31,9 @@ func getNonCanaryIngress(sources []providerir.BackendSource) *networkingv1.Ingre
 		}
 	}
 
+	if len(sources) > 0 {
+		return sources[0].Ingress
```
Comment (bexxmodd) [inline pkg/i2gw/providers/ingressnginx/utils.go:37]: Is it safe to always fallback to the first Ingress?

File: pkg/i2gw/providers/ingressnginx/backend_tls.go (hunk 16 above)
Comment (rajashish) [reply pkg/i2gw/providers/ingressnginx/backend_tls.go]: Thanks for catching that! I've updated the strict validation logic to require proxySSLServerName == "on". Since the default is "off", it must now be explicitly set in the Ingress annotations to comply with Gateway API's SNI requirement.

File: pkg/i2gw/providers/ingressnginx/utils.go (hunk 18 above)
Comment (rajashish) [reply pkg/i2gw/providers/ingressnginx/utils.go:37]: I've added a fallback comment to clarify this edge case. If all the ingress sources pointing to a standard backend somehow have a canary annotation (which shouldn't happen in a valid nginx setup since canaries derive from a primary), we fall back to the first available source to avoid a panic while parsing the common annotations

File: pkg/i2gw/providers/ingressnginx/backend_tls.go (hunk 16 above)
Comment (rajashish) [reply pkg/i2gw/providers/ingressnginx/backend_tls.go]: Thanks for catching that! I've updated the strict validation logic to require proxySSLServerName == "on". Since the default is "off", it must now be explicitly set in the Ingress annotations to comply with Gateway API's SNI requirement.

File: pkg/i2gw/providers/ingressnginx/backend_tls.go (hunk 17 above)
Comment (rajashish) [reply pkg/i2gw/providers/ingressnginx/backend_tls.go]: Good catch! You're completely right. I've removed that break statement. Now it correctly processes all backends within the rule so that setups like canary deployments map their policies properly to all targeted services.

File: pkg/i2gw/providers/ingressnginx/backend_tls.go (hunk 14 above)
Comment (rajashish) [reply pkg/i2gw/providers/ingressnginx/backend_tls.go]: done

File: pkg/i2gw/providers/ingressnginx/utils.go [hunk 19]
```diff
@@ -31,5 +31,11 @@ func getNonCanaryIngress(sources []providerir.BackendSource) *networkingv1.Ingre
 		}
 	}
 
+	// Fallback: If all sources are somehow canaries (invalid nginx config without a primary),
+	// we still return an Ingress to read annotations from to avoid panics.
+	if len(sources) > 0 {
+		return sources[0].Ingress
+	}
+
```
Comment (Stevenjin8) [inline pkg/i2gw/providers/ingressnginx/utils.go:39]:
```suggestion
```

**Comment by rajashish** [comment]
Hey @Stevenjin8 please review

**Comment by Stevenjin8** [comment]
@rajashish ci is not passing

**Comment by Stevenjin8** [comment]
@rajashish you also need to sign CLA

**Comment by bexxmodd** [comment]
`func backendTLSFeature` is convoluted beyond reasonable level, but I don't want to keep this PR on-hold. Hopefully, we can refactor this after release.

/lgtm
/approve

**Comment by Stevenjin8** [comment]
@rajashish Please sign CLA: https://github.com/kubernetes-sigs/ingress2gateway/pull/308#issuecomment-3771821450

**Comment by rajashish** [comment]
pull-ingress2gateway-test was failing. Fixed it. @bexxmodd please give LGTM again
